workflow/gccontroller/gc_controller.go

```diff
@@ -211,9 +211,24 @@ func (c *Controller) deleteWorkflow(ctx context.Context, key string) error {
 	// It should be impossible for a workflow to have been queue without a valid key.
 	namespace, name, _ := cache.SplitMetaNamespaceKey(key)
 
+	// Double check to fix retry-with-ttl bug (#12626).
+	wf, err := c.wfclientset.ArgoprojV1alpha1().Workflows(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if apierr.IsNotFound(err) {
+			log.Infof("Workflow already deleted '%s'", key)
+			return nil
+		} else {
+			return err
+		}
+	}
+	if !(wf.Status.Failed() || wf.Status.Successful() || wf.Status.Phase.Completed()) {
```

Comment:
TTL still take effect. because after workflow completed, it will insert ttl delaying queue again. More testing is needed to fully determine.

(However, retried WF may be deleted earlier than the TTL after WF completion. For example, the retried WF is completed quickly and  is not reached this branch that ignore deletion.)